lib/ReplaceOpenCLBuiltinPass.cpp

```diff
@@ -990,6 +992,15 @@ Value *ReplaceOpenCLBuiltinPass::replaceAsyncWorkGroupCopies(
   auto LocalSize1 = Builder.CreateExtractElement(LocalSize, Cst1);
   auto LocalSize2 = Builder.CreateExtractElement(LocalSize, Cst2);
 
+  if (clspv::PointersAre64Bit(M)) {
+    LocalId0 = Builder.CreateZExt(LocalId0, Builder.getInt64Ty());
+    LocalId1 = Builder.CreateZExt(LocalId1, Builder.getInt64Ty());
+    LocalId2 = Builder.CreateZExt(LocalId2, Builder.getInt64Ty());
+    LocalSize0 = Builder.CreateZExt(LocalSize0, Builder.getInt64Ty());
+    LocalSize1 = Builder.CreateZExt(LocalSize1, Builder.getInt64Ty());
+    LocalSize2 = Builder.CreateZExt(LocalSize2, Builder.getInt64Ty());
+  }
```

Comment:
I'm a bit confused as to why these are necessary.

Reply:
Here (and other places) the local id and local size are loaded directly from the global variables that contain them (`__spirv_WorkgroupSize` etc). Since these variables end up with the `WorkgroupSize` etc builtin decorations, they're always i32 even when targeting spir64. Generally I've tried to promote these to the index type as early as possible to avoid losing precision.

This could be replaced with calls to the builtins directly but after inlining the ZExts will be there anyway.